elitzur-avro/src/main/scala/com/spotify/elitzur/converters/avro/dynamic/DynamicAccessorCompanion.scala

```diff
@@ -0,0 +1,89 @@
+/*
+ * Copyright 2021 Spotify AB.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package com.spotify.elitzur.converters.avro.dynamic
+
+import com.spotify.elitzur.MetricsReporter
+import com.spotify.elitzur.converters.avro.AvroElitzurConversionUtils.byteBufferToByteArray
+import com.spotify.elitzur.validators.{
+  BaseCompanion,
+  BaseValidationType,
+  SimpleCompanionImplicit,
+  Validator
+}
+
+import scala.reflect.ClassTag
+import scala.reflect.runtime.universe.{TypeTag, typeOf}
+import java.{util => ju}
+import collection.JavaConverters._
+
+//scalastyle:off line.size.limit structural.type
+class DynamicAccessorCompanion[T: TypeTag, LT <: BaseValidationType[T]: ClassTag: ({type L[x] = SimpleCompanionImplicit[T, x]})#L] extends Serializable {
+//scalastyle:on line.size.limit structural.type
+
+  private val companion: BaseCompanion[T, LT] =
+    implicitly[SimpleCompanionImplicit[T, LT]].companion
+  private[dynamic] val validationType: String = companion.validationType
+
+  //scalastyle:off line.size.limit
+  private[dynamic] def getValidator(modifier: Modifier)(implicit metricReporter: MetricsReporter): Validator[Any] = {
+  //scalastyle:on line.size.limit
+    modifier match {
+      case Modifier.None => implicitly[Validator[LT]].asInstanceOf[Validator[Any]]
+      case Modifier.Opt => implicitly[Validator[Option[LT]]].asInstanceOf[Validator[Any]]
+      case Modifier.Seq => implicitly[Validator[Seq[LT]]].asInstanceOf[Validator[Any]]
+      case Modifier.SeqOpt => implicitly[Validator[Seq[Option[LT]]]].asInstanceOf[Validator[Any]]
+      case Modifier.OptSeq => implicitly[Validator[Option[Seq[LT]]]].asInstanceOf[Validator[Any]]
+      case Modifier.OptSeqOpt =>
+        implicitly[Validator[Option[Seq[Option[LT]]]]].asInstanceOf[Validator[Any]]
+    }
+  }
+
+  private[dynamic] def getPreprocessorForValidator(modifier: Modifier): Any => Any = {
+    def baseFn(v: Any): Any = parseAvro(v)
+    def toOptionFn(v: Any, fn: Any => Any): Any = Option(v).map(fn)
+    def toArrayFn(v: Any, fn: Any => Any): Any = v.asInstanceOf[ju.List[Any]].asScala.map(fn)
+
+    (fieldValue: Any) => modifier match {
+      case Modifier.None => baseFn(fieldValue)
```

Comment:
Would it be possible to do this recursively and just handle `None`, `Seq`, and `Opt`. Maybe if we had the Opt and Seq case objects store parameters representing type params?

If it's really complicated this should be fine for now though as I think it handles all the cases we care about

Reply:
I think it should be possible to make that work for `getValidator` as well but that might be a bit trickier

Reply:
I agree that the recursive approach for both is ideal (it certainly avoids us from having to worry about the different variations of `Modifiers` we want to support in the future).

I am looking into making `getValidator` recursive, but I think the dependence on implicitly identifying the class within each Validator is making things a bit tricky.